File: modules/custom/lua/ixi20_no_perpetuation.lua
-----------------------------------
-- Imagine XI 2.0: no avatar perpetuation MP cost (module only).
-- 1.0 returned 0 from PerpetuationCost(). Core src is left alone;
-- this zeros AVATAR_PERPETUATION after spawn / login. C++ module
-- also clears it on zone tick (level restriction recalcs the cost).
-- Gear with Avatar perpetuation cost is remapped in ixi20_no_perpetuation.sql.
-- Do not return this module.
-----------------------------------
require('modules/module_utils')
-----------------------------------

local m = Module:new('ixi20_no_perpetuation')

local function stripPerpetuation(player)
    if not player or not player.isPC or not player:isPC() then
        return
    end

    if player:getMod(xi.mod.AVATAR_PERPETUATION) ~= 0 then
        player:setMod(xi.mod.AVATAR_PERPETUATION, 0)
    end
end

m:addOverride('xi.pets.avatar.onMobSpawn', function(pet)
    super(pet)
    stripPerpetuation(pet:getMaster())
end)

m:addOverride('xi.pet.spawnPet', function(caster, petID, state, target)
    super(caster, petID, state, target)
    stripPerpetuation(caster)
end)

m:addOverride('xi.player.onGameIn', function(player, firstLogin, zoning)
    super(player, firstLogin, zoning)
    stripPerpetuation(player)
end)

File: modules/custom/cpp/ixi20_no_perpetuation.cpp
/************************************************************************
 * Imagine XI 2.0: no avatar perpetuation MP cost (module only).
 *
 * 1.0 made petutils::PerpetuationCost() return 0. This keeps core src
 * untouched and zeros AVATAR_PERPETUATION after spawn / zone / level
 * restriction recalcs. Lua ixi20_no_perpetuation.lua covers scripted
 * summons. Gear remaps live in ixi20_no_perpetuation.sql.
 ************************************************************************/

#include "map/utils/moduleutils.h"

#include "common/logging.h"

#include "map/entities/char_entity.h"
#include "map/modifier.h"
#include "map/zone.h"

namespace
{

void stripPerpetuation(CCharEntity* PChar)
{
    if (PChar == nullptr)
    {
        return;
    }

    if (PChar->getMod(xi::Mod::AVATAR_PERPETUATION) != 0)
    {
        PChar->setModifier(xi::Mod::AVATAR_PERPETUATION, 0);
    }
}

} // namespace

class Ixi20NoPerpetuationModule : public CPPModule
{
public:
    void OnInit() override
    {
        ShowInfo("Imagine XI 2.0: avatar perpetuation cost disabled");
    }

    void OnZoneTick(CZone* PZone) override
    {
        if (PZone == nullptr)
        {
            return;
        }

        PZone->ForEachChar([](CCharEntity* PChar) {
            stripPerpetuation(PChar);
        });
    }

    void OnCharZoneIn(CCharEntity* PChar) override
    {
        stripPerpetuation(PChar);
    }
};

REGISTER_CPP_MODULE(Ixi20NoPerpetuationModule);
